### PR TITLE
Restore unsaved buffers on restart

### PR DESCRIPTION
**Pull request by mrnugget**
This adds the ability for Zed to restore unsaved buffers on restart. The user is no longer prompted to save/discard/cancel when trying to close a Zed window with dirty buffers in it. Instead those dirty buffers are stored and restored on restart.

It does this by saving the contents of dirty buffers to the internal SQLite database in which Zed stores other data too. On restart, if there are dirty buffers in the database, they are restored.

On certain events (buffer changed, file saved, ...) Zed will serialize these buffers, throttled to a 100ms, so that we don't overload the machine by saving on every keystroke. When Zed quits, it waits until all the buffers are serialized.

### Current limitations
- It does not persist undo-history (right now we don't persist/restore undo-history regardless of dirty buffers or not)
- It does not restore buffers in windows without projects/worktrees. Example: if you open a new window with `cmd-shift-n` and type something in a buffer, this will _not_ be stored and you will be asked whether to save/discard on quit. In the future, we want to fix this by also restoring windows without projects/worktrees.

### Demo

https://github.com/user-attachments/assets/45c63237-8848-471f-8575-ac05496bba19

### Related tickets

I'm unsure about closing them, without also fixing the 2nd limitation: restoring of worktree-less windows. So let's wait until that.

- https://github.com/zed-industries/zed/issues/4985
- https://github.com/zed-industries/zed/issues/4683

### Note on performance

- Serializing editing buffer (asynchronously on background thread) with 500k lines takes ~200ms on M3 Max. That's an extreme case and that performance seems acceptable.

Release Notes:

- Added automatic restoring of unsaved buffers. Zed can now be closed even if there are unsaved changes in buffers. One current limitation is that this only works when having projects open, not single files or empty windows with unsaved buffers. The feature can be turned off by setting `{"session": {"restore_unsaved_buffers": false}}`.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -280,6 +280,7 @@ async-recursion = "1.0.0"
 async-tar = "0.4.2"
 async-trait = "0.1"
 async_zip = { version = "0.0.17", features = ["deflate", "deflate64"] }
+bincode = "1.2.1"
```
Comment (ConradIrwin) [inline Cargo.toml:283]: @mrnugget `bincode` doesn't have a good story around versioning - you can never change the struct definition that is serialized with it to add/remove fields for example. For something persisted to disk we should use json/bson or something that has a well-defined structure so it's harder to accidentally break from version to version.

Comment (mrnugget) [reply Cargo.toml:283]: Hehe, this was just temporary, we're not going to use `bincode` right now

Comment (ConradIrwin) [reply Cargo.toml]: Phew!

**Comment by mvp** [comment]
I have installed Zed 0.163.3 67d405ddf745641edff369ac2110f3f8864d6f8d, which seems quite recent.
However, restoring unsaved buffers on restart still does not seem to work.
I read the code, and it should be gated by setting `restore_unsaved_buffers` (defaults to `true`. I even added this setting explicitly to my config, but it still does not work.
Reading Zed book https://zed.dev/docs/configuring-zed, it does not mention this option at all.
What am I missing?
Basically, I am looking to be able to open few untitled/unsaved windows/scratchpads and have them survive restart or quitting the editor.

**Comment by ConradIrwin** [comment]
@mvp do you see any errors in your zed log? I've seen the sqlite database we use get corrupted occasionally

**Comment by mvp** [comment]
I cannot find any sqlite db under ~/.config/zed. Is there any other place?

Not supporting this transparently is very unfortunate.
Even if I am saving all files, quitting zed and starting it again (or rebooting my computer) doesn't load the same files again, which is even worse.

I guess back to Sublime Text :-(

**Comment by notpeter** [comment]
On MacOS the workspace is under ~/Library/Application Support/Zed

**Comment by mvp** [comment]
Opening a directory and then adding unsaved file works as expected, and it preserves last opened directory upon restart, as well does not ask to save unsaved files upon exit.

However, opening random files or scratches does not keep unsaved files, and does not remember list of previously opened files upon restart...

**Comment by mrnugget** [comment]
Yeah, that's a current limitation. See this note above in the release notes:

> One current limitation is that this only works when having projects open, not single files or empty windows with unsaved buffers.

**Comment by AndydeCleyre** [comment]
For completeness, the current tracking issue is https://github.com/zed-industries/zed/issues/15098

**Comment by babang1da** [comment]
dont work it! please fix
